File: algorithms/medium/longest_common_prefix_after_at_most_one_removal.py
class Solution:
    def longestCommonPrefix(self, s: str, t: str) -> int:
        N = len(s)
        M = len(t)
        skip = 0
        res = 0
        i, j = 0, 0
        while i < N and j < M:
            if s[i] != t[j]:
                if skip < 1:
                    skip += 1
                    i += 1
                else:
                    return res
            else:
                res += 1
                i += 1
                j += 1
        return res

# Main section
for s, t in [
               ('madxa', 'madam'),
               ('leetcode', 'eetcode'),
               ('one', 'one'),
               ('a', 'b'),
            ]:
    print(f's, t = {s}, {t}')
    sol = Solution()
    r = sol.longestCommonPrefix(s, t)
    print(f'r = {r}')
    print('========================')

File: algorithms/medium/longest_common_prefix_after_at_most_one_removal.go
package main

import "fmt"

func longestCommonPrefix(s string, t string) int {
	N := len(s)
	M := len(t)
	res := 0
	skip := 0
	i, j := 0, 0
	for i < N && j < M {
		if s[i] != t[j] {
			if skip < 1 {
				skip += 1
				i++
			} else {
				return res
			}
		} else {
			res++
			i++
			j++
		}
	}
	return res
}

func main() {
	data := [][]string{
		[]string{"madxa", "madam"},
		[]string{"leetcode", "eetcode"},
		[]string{"one", "one"},
		[]string{"a", "b"},
	}
	for _, v := range data {
		s, t := v[0], v[1]
		fmt.Printf("s, t = %s, %s\n", s, t)
		r := longestCommonPrefix(s, t)
		fmt.Printf("r = %d\n", r)
		fmt.Println("========================")
	}
}
